### PR TITLE
[Core] Fix bug in Parameters::Append(const Vector&)

### LINKED ISSUE #9079: Appending parameter arrays in python
**Description**
When attempting to append a ```Vector``` to a ```Parameters``` object that holds an array of ```Vector```s in python, only the last component is appended, creating a flat array instead of a vector array. Appending strings seem to work as intended, so it's probably a casting error in the python bindings, but I haven't found it yet.

**Scope**
Core

**To Reproduce**
```
import KratosMultiphysics

parameters = KratosMultiphysics.Parameters("""{
    "array_of_vectors" : [],
    "array_of_strings" : []
}""")

vector = KratosMultiphysics.Vector([1, 2, 3])
string = "abc"

for i in range(2):
    parameters["array_of_vectors"].Append(vector)
    parameters["array_of_strings"].Append(string)
    print(parameters)
```
Expecting
```
{
    "array_of_vectors" : [[1, 2, 3], [1, 2, 3]],
    "array_of_strings" : ["abc", "abc"]
}
```
but getting
```
{
    "array_of_vectors" : [3, 3],
    "array_of_strings" : ["abc", "abc"]
}
```
instead.

### PR DESCRIPTION
**Pull request by matekelemen**
**Description**
Fixes a bug where only the last component of a ```Vector``` was appended to a list ```Parameters``` object (closes #9079).

### CODE REVIEW

**Comment by matekelemen** [comment]
@philbucher
Two things I noticed during this PR:

- Since some tests in [```Parameters``` python test case](https://github.com/KratosMultiphysics/Kratos/blob/3507c438e45f47ddabd6cf4893a24cda8a8ce368/kratos/tests/test_kratos_parameters.py) unrelated to this PR fail but the CI isn't broken, I think they aren't executed (not added to any test suites). If this really is the case, I think we'd be better off with globbing for test cases at runtime to make sure every written test gets executed at some point in the pipeline, because I've seen (and made) this mistake before too. Sorting into test suites could be done through a settable flag in ```TestCase``` that has a default value for nightly builds for example. (I didn't add the tests to any test suites because that's out of this PR's scope.)

- Empty ```Matrix``` instances can have nonzero row sizes (as if matrices were stored as vectors of vectors internally). I guess this isn't that important and might be a feature of a third party library, but it's weird and might cause some minor performance loss nonetheless.

**Comment by philbucher** [comment]
> @philbucher
> Two things I noticed during this PR:
> 
> Since some tests in [`Parameters` python test case](https://github.com/KratosMultiphysics/Kratos/blob/3507c438e45f47ddabd6cf4893a24cda8a8ce368/kratos/tests/test_kratos_parameters.py) unrelated to this PR fail but the CI isn't broken, I think they aren't executed (not added to any test suites). If this really is the case, I think we'd be better off with globbing for test cases at runtime to make sure every written test gets executed at some point in the pipeline, because I've seen (and made) this mistake before too. Sorting into test suites could be done through a settable flag in `TestCase` that has a default value for nightly builds for example. (I didn't add the tests to any test suites because that's out of this PR's scope.)


Are you sure it isn't added? I think it is:
https://github.com/KratosMultiphysics/Kratos/blob/7fe497842aafbe6c9a5fbb6c4347ff70a5cfd276/kratos/tests/test_KratosCore.py#L89

I completely agree that globbing for them would be a better solution, but as you also pointed out that is not done bcs we use the different suites. (in my smaller project I use globbing and I think it is much better)
Do you know if there is a standard solution for this? Or is what you suggest a custom solution? Could you please elaborate a bit?

> Empty `Matrix` instances can have nonzero row sizes (as if matrices were stored as vectors of vectors internally). I guess this isn't that important and might be a feature of a third party library, but it's weird and might cause some minor performance loss nonetheless.

I am not sure I follow, can you give an example / explain in more detail?

Thx!

**Comment by matekelemen** [comment]
> Are you sure it isn't added? I think it is:

You're right, and now it passes for me too. Turns out that I have to run the tests from their containing directory or else a lot of them fail (not just IO issues).

> I completely agree that globbing for them would be a better solution, but as you also pointed out that is not done bcs we use the different suites. (in my smaller project I use globbing and I think it is much better)
Do you know if there is a standard solution for this? Or is what you suggest a custom solution? Could you please elaborate a bit?

https://github.com/matekelemen/auto_suite_example
There's a lot of boilerplate but the idea is dead simple:
- test [cases](https://github.com/matekelemen/auto_suite_example/blob/6173acddb1c53c39c83b625fd33806b443dcc363/CustomTestCase.py) and [suites](https://github.com/matekelemen/auto_suite_example/blob/6173acddb1c53c39c83b625fd33806b443dcc363/CustomTestSuite.py) have an extra member variable that describes which predefined suite they belong in. After [globbing](https://github.com/matekelemen/auto_suite_example/blob/6173acddb1c53c39c83b625fd33806b443dcc363/main.py#L20) test cases, they are [sorted into their respective suites](https://github.com/matekelemen/auto_suite_example/blob/6173acddb1c53c39c83b625fd33806b443dcc363/SuiteCollection.py#L13) and then run. Setting the system up such that each case can belong to multiple suites is trivial too.

> I am not sure I follow, can you give an example / explain in more detail?

```
import KratosMultiphysics

empty_matrix_0 = KratosMultiphysics.Matrix([[]])
print(empty_matrix_0)

empty_matrix_1 = KratosMultiphysics.Matrix([[],[]])
print(empty_matrix_1)
```
The size of ```empty_matrix_0``` is ```[0,0]``` as expected, but ```empty_matrix_1``` is of size ```[2,0]```.

File: kratos/sources/kratos_parameters.cpp [hunk 1]
```diff
@@ -954,7 +954,7 @@ void Parameters::Append(const Vector& rValue)
     nlohmann::json j_array(0.0, size);
 
     for (IndexType i = 0; i < size; ++i) {
-        j_array = rValue[i];
+        j_array[i] = rValue[i];
     }
```
Comment (tteschemacher) [inline kratos/sources/kratos_parameters.cpp]:
```suggestion
    j_array = rValue;
```

I think you can neglect the loop and can do just this.

Comment (matekelemen) [reply kratos/sources/kratos_parameters.cpp]: good catch!

**Comment by matekelemen** [comment]
Failures seem to be unrelated to this PR (#9092)

**Comment by roigcarlo** [comment]
Yep don't worry. Its from the nightly, no need to pass that to merge

**Review by tteschemacher — APPROVED** [review]
To me it looks fine now!

But maybe @philbucher wants to also leave a comment.

**Review by philbucher — APPROVED** [review]
Thanks!

**Comment by philbucher** [comment]
> > Are you sure it isn't added? I think it is:
> 
> You're right, and now it passes for me too. Turns out that I have to run the tests from their containing directory or else a lot of them fail (not just IO issues).
> 
> > I completely agree that globbing for them would be a better solution, but as you also pointed out that is not done bcs we use the different suites. (in my smaller project I use globbing and I think it is much better)
> > Do you know if there is a standard solution for this? Or is what you suggest a custom solution? Could you please elaborate a bit?
> 
> https://github.com/matekelemen/auto_suite_example
> There's a lot of boilerplate but the idea is dead simple:
> 
> test [cases](https://github.com/matekelemen/auto_suite_example/blob/6173acddb1c53c39c83b625fd33806b443dcc363/CustomTestCase.py) and [suites](https://github.com/matekelemen/auto_suite_example/blob/6173acddb1c53c39c83b625fd33806b443dcc363/CustomTestSuite.py) have an extra member variable that describes which predefined suite they belong in. After [globbing](https://github.com/matekelemen/auto_suite_example/blob/6173acddb1c53c39c83b625fd33806b443dcc363/main.py#L20) test cases, they are [sorted into their respective suites](https://github.com/matekelemen/auto_suite_example/blob/6173acddb1c53c39c83b625fd33806b443dcc363/SuiteCollection.py#L13) and then run. Setting the system up such that each case can belong to multiple suites is trivial too.
> 
> > I am not sure I follow, can you give an example / explain in more detail?
> 
> ```
> import KratosMultiphysics
> 
> empty_matrix_0 = KratosMultiphysics.Matrix([[]])
> print(empty_matrix_0)
> 
> empty_matrix_1 = KratosMultiphysics.Matrix([[],[]])
> print(empty_matrix_1)
> ```
> 
> The size of `empty_matrix_0` is `[0,0]` as expected, but `empty_matrix_1` is of size `[2,0]`.

@ashishdarekar since you recently implemented this, can you please take the time and fix it?
Thanks!

**Pull request merged**